docs/DefaultBundler.md

```diff
@@ -0,0 +1,376 @@
+
+We use the `stack` to track what bundleGroup we are currently in, so that we can accurately draw edges between bundles. `BundleIds` and `BundleGroupIds` are not the same, but can be. This is how we are able to reduce node bloat in the graph. In `createIdealGraph`, "bundleGroups" are simply represented by bundles that are attached to the root. Thus, when `bundleId === bundleGroupId`, or in this case, the tuple values are equivalent, we know we have entered a new bundleGroup.
+
+```js
+if (bundleIdTuple[0] === bundleIdTuple[1]) {
+  // Push to the stack on a new bundle group
+  stack.push([node.value, bundleIdTuple[0]]);
+}
+```
+
+The top of the stack will track what bundleGroup we are currently on.
+
+#### Code Split for Lazy and isolated Dependencies
+
+Consider the below code obtained from `integration/shared-bundle-single-source/index.js`
+
+<table><tr>
+<td>
+
+```js
+//index.js (project entry)
+import('./foo'); //async imports
+import('./bar'); //async
+```
+
+</td><td>
+
+```js
+//bar.js
+import styles from './styles.css';
+import html from './local.html'; //isolated
+```
+
+<td>
+</tr></table>
+
+![image info](./BundlerGraphs/steps/create-bundles-bundleGraph.png)
+
+IdealBundleGraph from `integration/shared-bundle-single-source/index.js`
+
+All the bundles above are explicit code-split points, but not all of them are "bundleGroups". As a rule of thumb, only Entry bundles, Async Bundles, and "isolated" Assets are bundleGroups. `Isolated` is found on the `bundleBehavior` property of bundles, which is typically set in a Transformer. `HTML` and images tend to be `isolated`.
+
+#### Code Split for Type Change, Parallel, and Inline Dependencies
+
+We also create bundles for type change assets, parallel dependencies, and inline assets.
+
+```js
+parentAsset.type !== childAsset.type ||
+  dependency.priority === 'parallel' ||
+  childAsset.bundleBehavior === 'inline';
+```
+
+Bundles created by this path are **not** bundleGroups, and can also be merged with bundles of the same time. The stack not only tracks what bundleGroup we're in, but also the last "parent" bundle, or `referencingBundleRoot`. We draw an edge from the `referencingBundleRoot` to our new bundle. We preemptively attempt to merge this bundle into any existing bundle of the same type **within our bundleGroup**. Merging is also handled in the next step, we track what bundles are eligible for merging via `typeChangeIds`.
+
+## Step: Merge Type Change Bundles
+
+Type change bundles are a special case of bundle, because they require consistent or “stable” names. As a result, we only allow one bundle of another type per bundleGroup. So, we need to merge bundles that exist within the same bundleGroups. (i.e. siblings)
```

Comment:
they only require stable names in entry bundle groups right?

Reply:
Ahh yes, you know what, I believe we used to have it such that it merged **any** bundles of the same type within a bundleGroup...but I see now we do `bundleGraph.nodes.entries` for the groups that are up for merging. Nice catch.

Updated it to reflect this